File: iFlat/web/app/model/bi/ProjectInfo.js
Ext.define('iFlat.model.bi.ProjectInfo', {
    extend: 'Ext.data.Model',
    fields: [
        {name: 'projectInfo.projNo', mapping: 'projNo', type: 'string'},
        {name: 'projectInfo.code', mapping: 'code', type: 'string'},
        {name: 'projectInfo.name', mapping: 'name', type: 'string'},
        {name: 'projectInfo.shortName', mapping: 'shortName', type: 'string'},
        {name: 'projectInfo.plannedPlace', mapping: 'plannedPlace', type: 'string'},
        {name: 'projectInfo.actualPlace', mapping: 'actualPlace', type: 'string'},
        {name: 'projectInfo.materialPct', mapping: 'materialPct', type: 'number', defaultValue: 65 },
        {name: 'projectInfo.manufacturingPct', mapping: 'manufacturingPct', type: 'number', defaultValue: 20 },
        {name: 'projectInfo.auxiliaryPct', mapping: 'auxiliaryPct', type: 'number', defaultValue: 10 },
        {name: 'projectInfo.owner', mapping: 'owner', type: 'string'},
        {name: 'projectInfo.surveyor', mapping: 'surveyor', type: 'string'},
        {name: 'projectInfo.deliveryDate', mapping: 'deliveryDate', type: 'date'},
        {name: 'projectInfo.amount', mapping: 'amount', type: 'number'},
        {name: 'projectInfo.currency', mapping: 'currency', type: 'string'},
        {name: 'projectInfo.contractRate', mapping: 'contractRate', type: 'number'},
        {name: 'projectInfo.usd', mapping: 'usd', type: 'number'},
        {name: 'projectInfo.actualRate', mapping: 'actualRate', type: 'number'},
        {name: 'projectInfo.cny', mapping: 'cny', type: 'number'},
        {name: 'projectInfo.usdAdd', mapping: 'usdAdd', type: 'number'},
        {name: 'projectInfo.cnyAdd', mapping: 'cnyAdd', type: 'number'},
        {name: 'projectInfo.commissionPct', mapping: 'commissionPct', type: 'number'},
        {name: 'projectInfo.commencePln', mapping: 'commencePln', type: 'date',},
        {name: 'projectInfo.commenceAct', mapping: 'commenceAct', type: 'date',},
        {name: 'projectInfo.shipwayPln', mapping: 'shipwayPln', type: 'date'},
        {name: 'projectInfo.shipwayAct', mapping: 'shipwayAct', type: 'date'},
        {name: 'projectInfo.launchPln', mapping: 'launchPln', type: 'date'},
        {name: 'projectInfo.launchAct', mapping: 'launchAct', type: 'date'},
        {name: 'projectInfo.seaTrialPln', mapping: 'seaTrialPln', type: 'date'},
        {name: 'projectInfo.seaTrialAct', mapping: 'seaTrialAct', type: 'date'},
        {name: 'projectInfo.deliveryPln', mapping: 'deliveryPln', type: 'date'},
        {name: 'projectInfo.deliveryAct', mapping: 'deliveryAct', type: 'date'},
        {name: 'projectInfo.blockPrdPln', mapping: 'blockPrdPln', type: 'number'},
        {name: 'projectInfo.blockPrdAct', mapping: 'blockPrdAct', type: 'number'},
        {name: 'projectInfo.shipwayPrdPln', mapping: 'shipwayPrdPln', type: 'number'},
        {name: 'projectInfo.shipwayPrdAct', mapping: 'shipwayPrdAct', type: 'number'},
        {name: 'projectInfo.dockPrdPln', mapping: 'dockPrdPln', type: 'number'},
        {name: 'projectInfo.dockPrdAct', mapping: 'dockPrdAct', type: 'number'},
        {name: 'projectInfo.buildPrdPln', mapping: 'buildPrdPln', type: 'number'},
        {name: 'projectInfo.buildPrdAct', mapping: 'buildPrdAct', type: 'number'},
        {name: 'projectInfo.analyseDate', mapping: 'analyseDate', type: 'date'},
    ]
});

File: src/main/java/com/iflat/bi/entity/ProjectInfo.java
package com.iflat.bi.entity;

import java.util.Date;

public class ProjectInfo {

    private String projNo;    //工号
    private String code;  //船号
    private String name;  //船名
    private String shortName;          //简称
    private String plannedPlace;       //计划船位
    private String actualPlace;        //实际船位
    private double materialPct;       //器材费指标
    private double manufacturingPct;  //加工费指标
    private double auxiliaryPct;      //专项费指标

    private String owner;       //船东
    private String surveyor;    //船检
    private Date deliveryDate;  //交船日
    private double amount;     //合同价
    private String currency;    //币种
    private double contractRate;  //合同汇率
    private double usd;        //结算美元
    private double actualRate; //实际汇率
    private double cny;        //结算人民币
    private double usdAdd;     //加帐美元
    private double cnyAdd;     //加帐人民币
    private double commissionPct; //佣金比例

    private Date commencePln;  //开工
    private Date shipwayPln;   //上船台
    private Date launchPln;    //下水
    private Date seaTrialPln;  //试航
    private Date deliveryPln;  //交付

    private Date commenceAct;  //开工
    private Date shipwayAct;   //上船台
    private Date launchAct;    //下水
    private Date seaTrialAct;  //试航
    private Date deliveryAct;  //交付

    private long blockPrdPln;   //分段周期
    private long shipwayPrdPln; //船台周期
    private long dockPrdPln;    //码头周期
    private long buildPrdPln;   //建造周期

    private long blockPrdAct;   //分段周期
    private long shipwayPrdAct; //船台周期
    private long dockPrdAct;    //码头周期
    private long buildPrdAct;   //建造周期

    private Date analyseDate;   //分析月份

    public Date getAnalyseDate() {
        return analyseDate;
    }

    public void setAnalyseDate(Date analyseDate) {
        this.analyseDate = analyseDate;
    }

    public String getProjNo() {
        return projNo;
    }

    public void setProjNo(String projNo) {
        this.projNo = projNo;
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getShortName() {
        return shortName;
    }

    public void setShortName(String shortName) {
        this.shortName = shortName;
    }

    public String getPlannedPlace() {
        return plannedPlace;
    }

    public void setPlannedPlace(String plannedPlace) {
        this.plannedPlace = plannedPlace;
    }

    public String getActualPlace() {
        return actualPlace;
    }

    public void setActualPlace(String actualPlace) {
        this.actualPlace = actualPlace;
    }

    public double getMaterialPct() {
        return materialPct;
    }

    public void setMaterialPct(double materialPct) {
        this.materialPct = materialPct;
    }

    public double getManufacturingPct() {
        return manufacturingPct;
    }

    public void setManufacturingPct(double manufacturingPct) {
        this.manufacturingPct = manufacturingPct;
    }

    public double getAuxiliaryPct() {
        return auxiliaryPct;
    }

    public void setAuxiliaryPct(double auxiliaryPct) {
        this.auxiliaryPct = auxiliaryPct;
    }

    public String getOwner() {
        return owner;
    }

    public void setOwner(String owner) {
        this.owner = owner;
    }

    public String getSurveyor() {
        return surveyor;
    }

    public void setSurveyor(String surveyor) {
        this.surveyor = surveyor;
    }

    public Date getDeliveryDate() {
        return deliveryDate;
    }

    public void setDeliveryDate(Date deliveryDate) {
        this.deliveryDate = deliveryDate;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public double getContractRate() {
        return contractRate;
    }

    public void setContractRate(double contractRate) {
        this.contractRate = contractRate;
    }

    public double getUsd() {
        return usd;
    }

    public void setUsd(double usd) {
        this.usd = usd;
    }

    public double getActualRate() {
        return actualRate;
    }

    public void setActualRate(double actualRate) {
        this.actualRate = actualRate;
    }

    public double getCny() {
        return cny;
    }

    public void setCny(double cny) {
        this.cny = cny;
    }

    public double getUsdAdd() {
        return usdAdd;
    }

    public void setUsdAdd(double usdAdd) {
        this.usdAdd = usdAdd;
    }

    public double getCnyAdd() {
        return cnyAdd;
    }

    public void setCnyAdd(double cnyAdd) {
        this.cnyAdd = cnyAdd;
    }

    public double getCommissionPct() {
        return commissionPct;
    }

    public void setCommissionPct(double commissionPct) {
        this.commissionPct = commissionPct;
    }

    public Date getCommencePln() {
        return commencePln;
    }

    public void setCommencePln(Date commencePln) {
        this.commencePln = commencePln;
        this.setBlockPrdPln();
        this.setBuildPrdPln();
    }

    public Date getShipwayPln() {
        return shipwayPln;
    }

    public void setShipwayPln(Date shipwayPln) {
        this.shipwayPln = shipwayPln;
        this.setBlockPrdPln();
        this.setShipwayPrdPln();
    }

    public Date getLaunchPln() {
        return launchPln;
    }

    public void setLaunchPln(Date launchPln) {
        this.launchPln = launchPln;
        this.setShipwayPrdPln();
        this.setDockPrdPln();
    }

    public Date getSeaTrialPln() {
        return seaTrialPln;
    }

    public void setSeaTrialPln(Date seaTrialPln) {
        this.seaTrialPln = seaTrialPln;
        this.setDockPrdPln();
    }

    public Date getDeliveryPln() {
        return deliveryPln;
    }

    public void setDeliveryPln(Date deliveryPln) {
        this.deliveryPln = deliveryPln;
        this.setBuildPrdPln();
    }

    public long getBlockPrdPln() {
        return blockPrdPln;
    }

    private void setBlockPrdPln() {
        this.blockPrdPln = periodCalc(this.commencePln, this.shipwayPln);
    }

    public long getShipwayPrdPln() {
        return shipwayPrdPln;
    }

    private void setShipwayPrdPln() {
        this.shipwayPrdPln = periodCalc(this.shipwayPln, this.launchPln);
    }

    public long getDockPrdPln() {
        return dockPrdPln;
    }

    private void setDockPrdPln() {
        this.dockPrdPln = periodCalc(this.launchPln, this.seaTrialPln);
    }

    public long getBuildPrdPln() {
        return buildPrdPln;
    }

    private void setBuildPrdPln() {
        this.buildPrdPln = periodCalc(this.commencePln, this.deliveryPln);
    }

    public Date getCommenceAct() {
        return commenceAct;
    }

    public void setCommenceAct(Date commenceAct) {
        this.commenceAct = commenceAct;
        this.setBlockPrdAct();
        this.setBuildPrdAct();
    }

    public Date getShipwayAct() {
        return shipwayAct;
    }

    public void setShipwayAct(Date shipwayAct) {
        this.shipwayAct = shipwayAct;
        this.setBlockPrdAct();
        this.setShipwayPrdAct();
    }

    public Date getLaunchAct() {
        return launchAct;
    }

    public void setLaunchAct(Date launchAct) {
        this.launchAct = launchAct;
        this.setShipwayPrdAct();
        this.setDockPrdAct();
    }

    public Date getSeaTrialAct() {
        return seaTrialAct;
    }

    public void setSeaTrialAct(Date seaTrialAct) {
        this.seaTrialAct = seaTrialAct;
        this.setDockPrdAct();
    }

    public Date getDeliveryAct() {
        return deliveryAct;
    }

    public void setDeliveryAct(Date deliveryAct) {
        this.deliveryAct = deliveryAct;
        this.setBuildPrdAct();
    }


    public long getBlockPrdAct() {
        return blockPrdAct;
    }

    private void setBlockPrdAct() {
        this.blockPrdAct = periodCalc(this.commenceAct, this.shipwayAct);
    }

    public long getShipwayPrdAct() {
        return shipwayPrdAct;
    }

    private void setShipwayPrdAct() {
        this.shipwayPrdAct = periodCalc(this.shipwayAct, this.launchAct);
    }

    public long getDockPrdAct() {
        return dockPrdAct;
    }

    private void setDockPrdAct() {
        this.dockPrdAct = periodCalc(this.launchAct, this.seaTrialAct);
    }

    public long getBuildPrdAct() {
        return buildPrdAct;
    }

    private void setBuildPrdAct() {
        this.buildPrdAct = periodCalc(this.commenceAct, this.deliveryAct);
    }

    private long periodCalc(Date start, Date end) {
        long diff = 0;
        if(start != null && end != null) {
            diff = end.getTime() - start.getTime();
        }
        return diff / (1000 * 60 * 60 * 24);
    }
}
